### PR TITLE
mrc-1642 more backend for sharing projects

### PR DESCRIPTION
**Pull request by hillalex**
This PR:
* adds an endpoint that will validate whether or not a provided email address exists
* modifies the clone project endpoint to accept a list of email addresses

### CODE REVIEW

File: src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt [hunk 1]
```diff
@@ -32,6 +32,7 @@ class MvcConfig(val config: Config) : WebMvcConfigurer {
         //Ajax endpoints only available to logged in users
         registry.addInterceptor(SecurityInterceptor(config, ""))
                 .addPathPatterns("/adr/**")
+                .addPathPatterns("/user/**")
```
Comment (hillalex) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt:35]: I noticed that the project related endpoints are not locked down to logged in users - is this intentional?

File: src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt [hunk 2]
```diff
@@ -32,6 +32,7 @@ class MvcConfig(val config: Config) : WebMvcConfigurer {
         //Ajax endpoints only available to logged in users
         registry.addInterceptor(SecurityInterceptor(config, ""))
                 .addPathPatterns("/adr/**")
+                .addPathPatterns("/user/**/*")
```
Comment (hillalex) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt:35]: I noticed that the project related endpoints are not locked down to logged in users - is this intentional?

File: src/app/src/test/kotlin/org/imperial/mrc/hint/integration/IntegrationTests.kt [hunk 3]
```diff
@@ -68,7 +68,7 @@ abstract class SecureIntegrationTests : CleanDatabaseTests() {
         when (isAuthorized) {
             IsAuthorized.TRUE -> {
                 Assertions.assertThat(responseEntity.headers.contentType!!.toString())
-                        .isEqualTo("application/json")
+                        .contains("application/json")
```
Comment (hillalex) [inline src/app/src/test/kotlin/org/imperial/mrc/hint/integration/IntegrationTests.kt:71]: When Spring wraps up the response entity it adds the charset UTF-8 bit to the end, unlike hintr, so just making this permissive of both.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt (hunk 1 above)
Comment (hillalex) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt:35]: I noticed that the project related endpoints are not locked down to logged in users - is this intentional?

**Comment by hillalex** [comment]
I can't work out for the life of me why these 2 tests are failing on Travis: https://travis-ci.com/github/mrc-ide/hint/jobs/392525054. They pass for me locally (and they follow a pattern that is working fine in other test classes.) Any ideas?!

**Comment by hillalex** [comment]
> I can't work out for the life of me why these 2 tests are failing on Travis: https://travis-ci.com/github/mrc-ide/hint/jobs/392525054. They pass for me locally (and they follow a pattern that is working fine in other test classes.) Any ideas?!

Ok, the plot thickens. When running the whole lot of tests locally (rather than an individual class) I *can* replicate this, but also all the ADR tests fail (the only other ones that are locked down to logged in users.) 

Something weird going on when running all tests at once:

* individually run all test classes pass
* as part of whole project test UserTests fail on Travis
* as part of whole project test UserTests ad ADRTests fail locally

**Comment by hillalex** [comment]
Again codecov complaining about a line that I believe is under test: https://github.com/mrc-ide/hint/pull/344/files#diff-67d5249d2948c462ff29af406d6a301dR172

File: src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt (hunk 1 above)
Comment (EmmaLRussell) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/AppStart.kt:35]: As discussed, no - other than the top level /projects endpoint, which should be open. I see you've changed the clone route now to `/project/{projectId}/clone` so maybe this is the place to add the interceptor for /project/**

File: src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ProjectTests.kt [hunk 4]
```diff
@@ -31,8 +28,7 @@ class ProjectTests : VersionFileTests() {
     private val testState = "{\"state\": \"test\"}"
 
     @Test
-    fun `can create new project`()
-    {
+    fun `can create new project`() {
```
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ProjectTests.kt:31]: Huh, all these methods have been changed to have opening { on same line as signature - I thought your preference was to put the opening { on a new line?

Comment (hillalex) [reply src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ProjectTests.kt:31]: As we don't have shared project style settings for hint, my IDE is using the kotlin defaults, which is brace on same line as here. We did discuss moving to using the more space-y style we use in montagu/OW, but that hasn't been implemented yet. I don't know why your IDE has formatted things with braces on next line unless you have that set as your default for kotlin?

Comment (hillalex) [reply src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ProjectTests.kt:31]: (nb I'm going to work on the style settings as part of "maintenance afternoon" this afternoon!)

**Comment by hillalex** [comment]
Re locking down project endpoints, it will require a bunch of changes to tests for other endpoints, so I was imagining that being done in one fell swoop as its own ticket.

**Review by EmmaLRussell — APPROVED** [review]

**Pull request merged**